### PR TITLE
Potential fix for HTML highlight issue

### PR DESCRIPTION
**Pull request by fnkabit**
Potential fix for issue #1933 
* fixes the issue at hand
* has some pros and cons compared to previous implementation
* **not properly tested** for all possible inputs

### CODE REVIEW

**Comment by pbek** [comment]
I love that you are working on it and am scared at the same time 😆

**Comment by fnkabit** [comment]
> I love that you are working on it and am scared at the same time

image

**Comment by pbek** [comment]
@Waqar144

**Comment by Waqar144** [comment]
This is mostly fine except for a couple of issues. The `=` in every attribute will get highlighted as "String", and the attribute needs to be `Format::Builtin`

File: src/helpers/codetohtmlconverter.cpp [hunk 1]
```diff
@@ -507,32 +508,30 @@ QString CodeToHtmlConverter::xmlHighlighter(const QStringRef &input) const {
...
+
+                while (matchIt.hasNext()) {
+                    QRegularExpressionMatch match = matchIt.next();
+                    QString captured = match.captured(0);
+
+                    if (!captured.contains(QLatin1Char('='))) {
+                        output += setFormat(QStringRef(&captured), Format::Keyword);
+                    } else {
+                        int eqPos = captured.indexOf(QLatin1Char('='));
+                        output += (setFormat(QStringRef(&captured, 0, eqPos), Format::Keyword) +
```
Comment (Waqar144) [inline src/helpers/codetohtmlconverter.cpp:523]: This should be `Format::Builtin`

File: src/helpers/codetohtmlconverter.cpp [hunk 2]
```diff
@@ -507,32 +508,30 @@ QString CodeToHtmlConverter::xmlHighlighter(const QStringRef &input) const {
...
+                                   setFormat(QStringRef(&captured, eqPos, captured.length() - eqPos), Format::String));
```
Comment (Waqar144) [inline src/helpers/codetohtmlconverter.cpp:524]: This will include the `=` as well when highlighting the string

**Review by Waqar144 — COMMENTED** [review]
The next `else if (input.at(i) == QLatin1Char('=')) ` is unnecessary with these changes

**Comment by Waqar144** [comment]
This can be handled in a couple of other ways, i.e create a nested loop so we can handle all the attributes in one go quickly. Or we can just put a condition in the `else` to see if we should include the current character in output or not. Something like:

```
else {
            int equalPos = input.indexOf(QLatin1Char('='), i);
            if (equalPos != -1) {
                int newline = input.indexOf(QLatin1Char('\n'), i);
                if (newline > equalPos) {
                    i = equalPos - 1;
                    continue;
                }
// include in output
            output += escape(input.at(i));
            }
```

**Comment by fnkabit** [comment]
Thank you for the review @Waqar144 . I modified the PR accordingly.

**Review by Waqar144 — APPROVED** [review]

**Comment by Waqar144** [comment]
Thanks @fnkabit, Looks awesome.

**Comment by pbek** [comment]
Great, thank you!

**Comment by pbek** [comment]
@fnkabit, @Waqar144, when I drop the demo markdown from the issue into a note QON crashes...

    ```html
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    ```
    
    ```js
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    ```


```
__GI_raise 0x00007f58d974418b
__GI_abort 0x00007f58d9723859
QMessageLogger::fatal(char const*, ...) const 0x00007f58d9ccfc39
qt_assert(char const*, char const*, int) 0x00007f58d9ccefe3
QString::at qstring.h:1067
QStringRef::at qstring.h:1748
CodeToHtmlConverter::escapeString codetohtmlconverter.cpp:779
CodeToHtmlConverter::setFormat codetohtmlconverter.cpp:798
CodeToHtmlConverter::xmlHighlighter codetohtmlconverter.cpp:525
CodeToHtmlConverter::process codetohtmlconverter.cpp:116
highlightCode note.cpp:2239
Note::textToMarkdownHtml note.cpp:2319
Note::toMarkdownHtml note.cpp:2190
MainWindow::setNoteTextFromNote mainwindow.cpp:4422
MainWindow::noteViewUpdateTimerSlot mainwindow.cpp:2941
QtPrivate::FunctorCall<QtPrivate::IndexesList<>, QtPrivate::List<>, void, void (MainWindow::*)()>::call(void (MainWindow::*)(), MainWindow*, void**) qobjectdefs_impl.h:152
QtPrivate::FunctionPointer<void (MainWindow::*)()>::call<QtPrivate::List<>, void>(void (MainWindow::*)(), MainWindow*, void**) qobjectdefs_impl.h:185
QtPrivate::QSlotObject<void (MainWindow::*)(), QtPrivate::List<>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) qobjectdefs_impl.h:418
<unknown> 0x00007f58d9f245ae
QTimer::timeout(QTimer::QPrivateSignal) 0x00007f58d9f290de
QObject::event(QEvent*) 0x00007f58d9f1b045
QApplicationPrivate::notify_helper(QObject*, QEvent*) 0x00007f58dabf0cc3
QApplication::notify(QObject*, QEvent*) 0x00007f58dabf9c70
QCoreApplication::notifyInternal2(QObject*, QEvent*) 0x00007f58d9eed6aa
QTimerInfoList::activateTimers() 0x00007f58d9f471e0
<unknown> 0x00007f58d9f47acc
g_main_context_dispatch 0x00007f58d8dc1fbd
<unknown> 0x00007f58d8dc2240
g_main_context_iteration 0x00007f58d8dc22e3
QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) 0x00007f58d9f47e92
QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) 0x00007f58d9eec1bb
QCoreApplication::exec() 0x00007f58d9ef4354
main main.cpp:593
__libc_start_main 0x00007f58d97250b3
_start 0x00005651fe58181e
```

**Comment by pbek** [comment]
Actually, even an empty html code block is enough to crash the app...

**Comment by fnkabit** [comment]
Don't know why I'm not getting the crash but I think it's because I'm off by 1.

**Comment by fnkabit** [comment]
@pbek Will create a PR asap.

**Comment by pbek** [comment]
I tried it out and it worked.
But now QON crashes every time I use the keyboard cursors in a note!?

**Comment by fnkabit** [comment]
> I tried it out and it worked.
> But now QON crashes every time I use the keyboard cursors in a note!?

Do you have a stack trace, please ?

**Comment by pbek** [comment]
Must have been something with my debugger. Tried with clion and it worked there...

**Comment by pbek** [comment]
Managed to crash it again... strange

```
1   QMarkdownTextEdit::eventFilter                                                                       qmarkdowntextedit.cpp         288 0x5555558ac5b7 
2   QOwnNotesMarkdownTextEdit::eventFilter                                                               qownnotesmarkdowntextedit.cpp 703 0x5555557e5fc1 
3   QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject *, QEvent *)                                                            0x7ffff624d413 
4   QApplicationPrivate::notify_helper(QObject *, QEvent *)                                                                                0x7ffff78d2cb2 
5   QApplication::notify(QObject *, QEvent *)                                                                                              0x7ffff78dc3a6 
6   QCoreApplication::notifyInternal2(QObject *, QEvent *)                                                                                 0x7ffff624d6aa 
7   ??                                                                                                                                     0x7ffff7934506 
8   QApplicationPrivate::notify_helper(QObject *, QEvent *)                                                                                0x7ffff78d2cc3 
9   QApplication::notify(QObject *, QEvent *)                                                                                              0x7ffff78dbc70 
10  QCoreApplication::notifyInternal2(QObject *, QEvent *)                                                                                 0x7ffff624d6aa 
11  QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent *)                                                     0x7ffff71be3eb 
12  QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent *)                                   0x7ffff71c359b 
13  QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>)                                                  0x7ffff719b2cb 
14  ??                                                                                                                                     0x7ffff2b1099e 
15  g_main_context_dispatch                                                                                                                0x7ffff50fefbd 
16  ??                                                                                                                                     0x7ffff50ff240 
17  g_main_context_iteration                                                                                                               0x7ffff50ff2e3 
18  QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>)                                                             0x7ffff62a7e92 
19  QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>)                                                                                0x7ffff624c1bb 
20  QCoreApplication::exec()                                                                                                               0x7ffff6254354 
... <More>                                                                                                                                                
```

**Comment by pbek** [comment]
Strange, ignore it for now. Must be my Qt Creator...

**Comment by Waqar144** [comment]
> Having some issues compiling in Debug mode; so I am unable to test this, but I think it's because I'm off by 1 at `codetohtmlconverter.cpp:525`.
> It should be `setFormat(QStringRef(&captured, eqPos + 1, captured.length() - eqPos - 1), Format::String));`

Crash could be here.  `eqPos + 1` can be out of bounds

**Comment by pbek** [comment]
So you mean https://github.com/pbek/QOwnNotes/commit/a2895325cbc53dd83b0a413a8ed8159622f1c2dc is not enough?
I wasn't able to crash it...

**Comment by Waqar144** [comment]
I think it will do, we'll see if anything breaks. Haven't tested it myself yet. Is the highlighting okay?

**Comment by pbek** [comment]
> Is the highlighting okay?

Yes, it seems!

**Pull request merged**